### PR TITLE
Exploration for a data layer

### PR DESCRIPTION
**Pull request by altendky**
Draft for:
- [ ] Being more than an exploration
- [ ] Moving to a new data layer repository

### CODE REVIEW

**Comment by altendky** [comment]
Alrighty.  Should I look to https://github.com/chia-network/chiapos for layout and CI reference?

A quick grep didn't turn up any imports of `chia` in `chiapos`, but I already am referencing some tooling located in chia-blockchain here such as `chia.types.blockchain_format.sized_bytes.bytes32` and `chia.util.db_wrapper.DBWrapper`.  Are we just ok with bidirectional dependencies between repos, or is this a sign I'm doing something wrong?  I suppose really that it's unlikely to be exclusively 'here' or 'there' but rather both in which case maybe it's ok to develop here until we settle on what code belongs where and deal with the extra CI work then?

Or, I may just be thinking about things entirely aside from your reasons.  I'm all for granular repositories.  That often leads to breaking other things down as well like putting utilities in a 'third' project which has corresponding overhead either in more pre-design to avoid frequent updates or just dealing with frequent updates and compatibility.  I have no history here to know how those decisions have been or will be made.

**Comment by hoffmang9** [comment]
I think the implied process being proposed here is to make chia-blockchain a sub repo of a datalayer repo.

**Comment by altendky** [comment]
Hmm, ok.  I was thinking of data layer as another service to run but that sounds more like a separate application that happens to use some existing bits in `chia-blockchain`.  I'll adjust my perspective.

Perhaps it is still ok to leave the extraction and CI setup for a bit later.  I noted it in the OP.

**Comment by mariano54** [comment]
The data layer is a service, kind of like the pool is a service, and layer two lighning nodes will be services. However, they are not part of the core blockchain (layer 1), and therefore should be a new repo.
I like the structure of using chia-blockchain by including a subrepo like Gene said, and then requiring the node to be running in order to be able to run the service.
One key difference with these layer 2 services (vs services in chia-blockchain) is that they don't use the Chia protocol (websockets, handshake, node ids, etc), and instead just use the RPC, since they are most separated. This keeps the code simpler.

**Comment by altendky** [comment]
Thanks for the explanation.  I'll take a look at https://github.com/chia-network/pool-reference for familiarity.

**Comment by hoffmang9** [comment]
@cmmarslender pointed out a potentially much better way to depend on chia-blockchain as he was cleaning up chiadns. Instead of submoduling chia-blockchain, just make it a dependency in setup.py for the datalayer.

**Comment by mariano54** [comment]
> @cmmarslender pointed out a potentially much better way to depend on chia-blockchain as he was cleaning up chiadns. Instead of submoduling chia-blockchain, just make it a dependency in setup.py for the datalayer.

oh you're right! That is better. That's what I meant, it's how pool reference works.

File: chia/data_layer/data_store.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        )
+        # TODO: Could also be structured such that the action table has a reference from
+        #       each action to the commit it is part of.
+        await self.db.execute("CREATE TABLE IF NOT EXISTS commits(" "changelist_hash TEXT, actions_index INTEGER" ")")
+
+        await self.db.commit()
+
+        return self
+
+    # TODO: Add some handling for multiple tables.  Could be another layer of class
```
Comment (almogdepaz) [inline chia/data_layer/data_store.py:99]: we can make each row can contain a "Table" column containing the id of the table it belongs to.
that way we can manage adding an removing tables easy

Comment (altendky) [reply chia/data_layer/data_store.py:101]: Agreed, I meant multiple data layer tables, not multiple sql tables.

File: chia/rpc/rpc_server.py [hunk 2]
```diff
@@ -101,14 +101,14 @@ async def get_connections(self, request: Dict) -> Dict:
         request_node_type: Optional[NodeType] = None
         if "node_type" in request:
             request_node_type = NodeType(request["node_type"])
-        if self.rpc_api.service.server is None:
+        if self.rpc_api.data_layer_service.server is None:
```
Comment (altendky) [inline chia/rpc/rpc_server.py:104]: @almogdepaz, it looks like maybe a refactor rename reached a bit far out?  Though I think maybe this code expects the RPC implementations to use `.service`?

Comment (altendky) [reply chia/rpc/rpc_server.py:104]: I suppose I ought to link to an example error in the CI runs.

https://github.com/Chia-Network/chia-blockchain/pull/8492/checks?check_run_id=3646582478#step:14:729
```
rpc_api = <chia.rpc.full_node_rpc_api.FullNodeRpcApi object at 0x7f0400847370>
self_hostname = 'localhost', daemon_port = 55400, rpc_port = 51000
stop_cb = <bound method Service.stop of <chia.server.start_service.Service object at 0x7f0401b15220>>
root_path = PosixPath('/tmp/tmpt5azhi35')
net_config = {'ALERTS_URL': 'download.chia.net/notify/mainnet_alert.txt', 'CHIA_ALERTS_PUBKEY': '89b7fd87cb56e926ecefb879a2...a': {'crt': 'config/ssl/ca/chia_ca.crt', 'key': 'config/ssl/ca/chia_ca.key'}, 'daemon_max_message_size': 50000000, ...}
connect_to_daemon = False

    async def start_rpc_server(
        rpc_api: Any,
        self_hostname: str,
        daemon_port: uint16,
        rpc_port: uint16,
        stop_cb: Callable,
        root_path: Path,
        net_config,
        connect_to_daemon=True,
    ):
        """
        Starts an HTTP server with the following RPC methods, to be used by local clients to
        query the node.
        """
        app = aiohttp.web.Application()
        rpc_server = RpcServer(rpc_api, rpc_api.service_name, stop_cb, root_path, net_config)
>       rpc_server.rpc_api.data_layer_service._set_state_changed_callback(rpc_server.state_changed)
E       AttributeError: 'FullNodeRpcApi' object has no attribute 'data_layer_service'
```

File: chia/data_layer/data_store.py [hunk 3]
```diff
@@ -108,29 +109,68 @@ async def get_row_by_index(self, index: int) -> CLVMObject:
     async def get_row_by_hash(self, row_hash: bytes32) -> CLVMObject:
         pass
 
-    async def insert_row(self, table_id: bytes32, clvm_object: CLVMObject) -> None:
+    async def insert_row(self, table: bytes32, clvm_object: CLVMObject, index: Optional[int] = None) -> None:
```
Comment (almogdepaz) [inline chia/data_layer/data_store.py:112]: does the user determine the row index  on submission ? 
i think that it should just be a running id

Comment (altendky) [reply chia/data_layer/data_store.py:112]: All the examples had `Row #1` etc so I added indexes.  As I look back, I may well have been reading too much into that and it may be removed.  It is a point to discuss as to whether the rows have order to them.

File: chia/cmds/chia.py [hunk 4]
```diff
@@ -118,6 +119,7 @@ def run_daemon_cmd(ctx: click.Context, wait_for_unlock: bool) -> None:
 cli.add_command(stop_cmd)
 cli.add_command(netspace_cmd)
 cli.add_command(farm_cmd)
+cli.add_command(data_cmd)
```
Comment (almogdepaz) [inline chia/cmds/chia.py:123]: ok but seems that when we split to a different repo this should change and the main chia node should not include commands related to data-layer

File: chia/rpc/rpc_server.py [hunk 5]
```diff
@@ -302,40 +302,54 @@ async def start_rpc_server(
     Starts an HTTP server with the following RPC methods, to be used by local clients to
     query the node.
     """
```
Comment (almogdepaz) [inline chia/rpc/rpc_server.py:304]: this refactor seems unrelated to the data-layer, i suggest you cherry pick it into another pr

Comment (altendky) [reply chia/rpc/rpc_server.py:304]: (discussed elsewhere, for the record here...)  It certainly is, it was added to identify a problem I ran into here so I kept it around.  But yes, if committed to chia-blockchain it would be via a different PR for just this.

File: chia/data_layer/data_store.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+        self.db.row_factory = aiosqlite.Row
+
+        # TODO: what pragmas do we want?  maybe foreign_keys?
+        # await self.db.execute("pragma journal_mode=wal")
+        # await self.db.execute("pragma synchronous=2")
+
+        # TODO: make this handle multiple data layer tables
+        # TODO: do we need to handle multiple equal rows
+
+        await self.db.execute("CREATE TABLE IF NOT EXISTS tables(id TEXT PRIMARY KEY, name STRING)")
```
Comment (almogdepaz) [inline chia/data_layer/data_store.py:147]: they does a table need two identifiers (name and id ?)

Comment (altendky) [reply chia/data_layer/data_store.py:123]: No, that's just me going beyond where I should and presuming that a human readable name would be good someplace.  Though, it probably belongs on the NFT or such if anywhere.
